File: tools/python/reroom_verify/wire_frame.py
"""Strict RRFP-WIRE-1 encoder and decoder."""

from __future__ import annotations

import hashlib
import struct
from typing import Any

from .canonical_json import artifact, canonical_bytes
from .loader import LoadedFixture, VerificationFailure, parse_json_bytes
from .schema_validator import _validate_schema


MAGIC = b"RRFP"
VERSION_MAJOR = 1
VERSION_MINOR = 0
FIXED_HEADER_BYTES = 24
MAX_JSON_HEADER_BYTES = 65_536
MAX_PAYLOAD_BYTES = 16_777_216
HEADER_STRUCT = struct.Struct(">4sBBHIIQ")


def encode_frame(header: dict[str, Any], payload: bytes) -> bytes:
    header_bytes = canonical_bytes(header)
    if len(header_bytes) > MAX_JSON_HEADER_BYTES or len(payload) > MAX_PAYLOAD_BYTES:
        raise VerificationFailure("wire_length", "wire field exceeds its bound")
    declared_length = header.get("image", {}).get("payload", {}).get("byte_length")
    if declared_length != len(payload):
        raise VerificationFailure("wire_length", "payload length duplicate mismatch")
    if header.get("payload_sha256") != hashlib.sha256(payload).hexdigest():
        raise VerificationFailure("digest_mismatch", "payload digest mismatch")
    sequence = header.get("capture_sequence")
    if not isinstance(sequence, int) or isinstance(sequence, bool) or not 0 <= sequence < 2**64:
        raise VerificationFailure("wire_sequence", "capture sequence is outside uint64")
    fixed = HEADER_STRUCT.pack(
        MAGIC,
        VERSION_MAJOR,
        VERSION_MINOR,
        0,
        len(header_bytes),
        len(payload),
        sequence,
    )
    return fixed + header_bytes + payload


def decode_frame(frame: bytes, *, fixture: LoadedFixture | None = None) -> tuple[dict[str, Any], bytes]:
    if len(frame) < FIXED_HEADER_BYTES:
        raise VerificationFailure("wire_truncated", "fixed header is truncated")
    magic, major, minor, flags, header_length, payload_length, sequence = HEADER_STRUCT.unpack_from(frame)
    if magic != MAGIC:
        raise VerificationFailure("wire_magic", "wire magic mismatch")
    if (major, minor) != (VERSION_MAJOR, VERSION_MINOR):
        raise VerificationFailure("wire_version", "wire version mismatch")
    if flags != 0:
        raise VerificationFailure("wire_flags", "reserved flags must be zero")
    if not 0 < header_length <= MAX_JSON_HEADER_BYTES:
        raise VerificationFailure("wire_length", "JSON header length is invalid")
    if payload_length > MAX_PAYLOAD_BYTES:
        raise VerificationFailure("wire_length", "payload length is invalid")
    expected_length = FIXED_HEADER_BYTES + header_length + payload_length
    if len(frame) < expected_length:
        raise VerificationFailure("wire_truncated", "wire frame is truncated")

    header_bytes = frame[FIXED_HEADER_BYTES : FIXED_HEADER_BYTES + header_length]
    try:
        header = parse_json_bytes(header_bytes)
    except VerificationFailure as error:
        raise VerificationFailure("wire_length", "declared JSON header boundary is invalid") from error
    if not isinstance(header, dict) or canonical_bytes(header) != header_bytes:
        raise VerificationFailure("schema_validation", "wire JSON header is not canonical")
    if fixture is not None:
        _validate_schema(fixture, "docs/contracts/frame-packet.schema.json", header)
    declared_payload_length = header.get("image", {}).get("payload", {}).get("byte_length")
    if declared_payload_length != payload_length:
        raise VerificationFailure("wire_length", "payload length duplicate mismatch")
    if header.get("capture_sequence") != sequence:
        raise VerificationFailure("wire_sequence", "capture sequence duplicate mismatch")
    payload_start = FIXED_HEADER_BYTES + header_length
    payload = frame[payload_start:expected_length]
    if header.get("payload_sha256") != hashlib.sha256(payload).hexdigest():
        raise VerificationFailure("digest_mismatch", "payload digest mismatch")
    if len(frame) > expected_length:
        raise VerificationFailure("wire_trailing_bytes", "wire frame has trailing bytes")
    return header, payload


def _apply_wire_mutations(frame: bytes, mutations: list[dict[str, Any]]) -> bytes:
    value = bytearray(frame)
    for mutation in mutations:
        operation = mutation["op"]
        if operation == "replace_byte":
            offset = mutation["offset"]
            replacement = bytes.fromhex(mutation["value_hex"])
            if len(replacement) != 1 or not 0 <= offset < len(value):
                raise VerificationFailure("semantic_invariant", "invalid byte mutation")
            value[offset] = replacement[0]
        elif operation == "replace_u32be":
            struct.pack_into(">I", value, mutation["offset"], mutation["value"])
        elif operation == "replace_u64be":
            struct.pack_into(">Q", value, mutation["offset"], mutation["value"])
        elif operation == "append_hex":
            value.extend(bytes.fromhex(mutation["value_hex"]))
        elif operation == "truncate":
            value = value[: mutation["byte_length"]]
        else:
            raise VerificationFailure("semantic_invariant", "unknown wire mutation")
    return bytes(value)


def _load_base_hex(fixture: LoadedFixture, relative_path: str) -> bytes:
    try:
        text = fixture.read_fixture_file(relative_path).decode("ascii").strip()
        return bytes.fromhex(text)
    except (UnicodeDecodeError, ValueError) as error:
        raise VerificationFailure("json_parse", "wire oracle is not hexadecimal") from error


def execute_wire_case(fixture: LoadedFixture, case: dict[str, Any]) -> list[dict[str, Any]]:
    spec = parse_json_bytes(
        fixture.read_fixture_file(case["input"]["relative_path"]),
        max_depth=fixture.max_document_depth,
    )
    if case["case_kind"] == "wire_bytes":
        header_path = (fixture.fixture_root / spec["header_source"]).resolve(strict=True)
        if not header_path.is_relative_to(fixture.repo_root):
            raise VerificationFailure("invalid_path", "wire header source escapes repository")
        header = parse_json_bytes(header_path.read_bytes(), max_depth=fixture.max_document_depth)
        try:
            payload = bytes.fromhex(spec["payload_hex"])
        except ValueError as error:
            raise VerificationFailure("json_parse", "wire payload is not hexadecimal") from error
        frame = encode_frame(header, payload)
        decode_frame(frame, fixture=fixture)
        return [artifact("wire_bytes", (frame.hex() + "\n").encode("ascii"))]
    if case["case_kind"] == "wire_mutation":
        frame = _load_base_hex(fixture, spec["base"])
        mutated = _apply_wire_mutations(frame, spec["mutations"])
        decode_frame(mutated, fixture=fixture)
        return []
    raise VerificationFailure("semantic_invariant", "unknown wire case kind")

File: packages/contracts/src/wire-frame.mjs
import { canonicalizeBytes } from "./canonical-json.mjs";
import {
  caseResult,
  parseJsonBytesStrict,
  RunnerFailure,
  readFixtureFile,
  readFixtureInternal,
  rejectedCaseResult,
  sha256Hex,
} from "./loader.mjs";
import { validateContractValue } from "./schema-validator.mjs";

export const RRFP_FIXED_HEADER_BYTES = 24;
export const RRFP_JSON_MAX_BYTES = 65_536;
export const RRFP_PAYLOAD_MAX_BYTES = 16_777_216;

function requireWire(condition, rejectionClass, message) {
  if (!condition) throw new RunnerFailure(rejectionClass, message);
}

function requireSequence(value) {
  requireWire(
    Number.isSafeInteger(value) && value >= 0,
    "wire_sequence",
    "capture sequence is not an exact nonnegative integer",
  );
  return BigInt(value);
}

export function encodeWireFrame(header, payloadBytes) {
  const payload = Buffer.from(payloadBytes);
  const headerBytes = canonicalizeBytes(header);
  requireWire(
    headerBytes.length <= RRFP_JSON_MAX_BYTES,
    "wire_length",
    "JCS header exceeds RRFP limit",
  );
  requireWire(
    payload.length <= RRFP_PAYLOAD_MAX_BYTES,
    "wire_length",
    "payload exceeds RRFP limit",
  );
  requireWire(
    header.image?.payload?.byte_length === payload.length,
    "wire_length",
    "payload length duplicate does not agree",
  );
  requireWire(
    header.payload_sha256 === sha256Hex(payload),
    "digest_mismatch",
    "payload digest does not agree",
  );
  const fixed = Buffer.alloc(RRFP_FIXED_HEADER_BYTES);
  fixed.write("RRFP", 0, "ascii");
  fixed.writeUInt8(1, 4);
  fixed.writeUInt8(0, 5);
  fixed.writeUInt16BE(0, 6);
  fixed.writeUInt32BE(headerBytes.length, 8);
  fixed.writeUInt32BE(payload.length, 12);
  fixed.writeBigUInt64BE(requireSequence(header.capture_sequence), 16);
  return Buffer.concat([fixed, headerBytes, payload]);
}

export function decodeWireFrame(fixture, bytes) {
  const wire = Buffer.from(bytes);
  requireWire(
    wire.length >= RRFP_FIXED_HEADER_BYTES,
    "wire_truncated",
    "RRFP fixed header is truncated",
  );
  requireWire(
    wire.subarray(0, 4).equals(Buffer.from("RRFP", "ascii")),
    "wire_magic",
    "RRFP magic is invalid",
  );
  requireWire(
    wire.readUInt8(4) === 1 && wire.readUInt8(5) === 0,
    "wire_version",
    "RRFP version is unsupported",
  );
  requireWire(wire.readUInt16BE(6) === 0, "wire_flags", "RRFP flags must be zero");
  const headerLength = wire.readUInt32BE(8);
  const payloadLength = wire.readUInt32BE(12);
  const sequence = wire.readBigUInt64BE(16);
  requireWire(
    headerLength <= RRFP_JSON_MAX_BYTES && payloadLength <= RRFP_PAYLOAD_MAX_BYTES,
    "wire_length",
    "RRFP declared length exceeds limit",
  );
  requireWire(
    wire.length >= RRFP_FIXED_HEADER_BYTES + headerLength,
    "wire_truncated",
    "RRFP JSON header is truncated",
  );

  const headerBytes = wire.subarray(
    RRFP_FIXED_HEADER_BYTES,
    RRFP_FIXED_HEADER_BYTES + headerLength,
  );
  let header;
  try {
    header = parseJsonBytesStrict(headerBytes);
  } catch (error) {
    if (error instanceof RunnerFailure)
      throw new RunnerFailure(
        "wire_length",
        "RRFP JSON-header length does not delimit one complete header",
      );
    throw error;
  }
  validateContractValue(fixture, "CON-001", header);
  requireWire(
    canonicalizeBytes(header).equals(headerBytes),
    "wire_length",
    "RRFP JSON header is not exact RFC 8785 JCS bytes",
  );
  requireWire(
    header.image?.payload?.byte_length === payloadLength,
    "wire_length",
    "RRFP payload-length duplicates disagree",
  );
  requireWire(
    requireSequence(header.capture_sequence) === sequence,
    "wire_sequence",
    "RRFP capture-sequence duplicates disagree",
  );

  const expectedLength = RRFP_FIXED_HEADER_BYTES + headerLength + payloadLength;
  requireWire(wire.length >= expectedLength, "wire_truncated", "RRFP payload is truncated");
  requireWire(wire.length === expectedLength, "wire_trailing_bytes", "RRFP has trailing bytes");
  const payload = wire.subarray(RRFP_FIXED_HEADER_BYTES + headerLength);
  requireWire(
    header.payload_sha256 === sha256Hex(payload),
    "digest_mismatch",
    "RRFP payload SHA-256 mismatch",
  );
  return { header, payload };
}

function parseWireHex(bytes) {
  const text = bytes.toString("utf8");
  requireWire(
    /^[0-9a-f]+\n$/.test(text) && (text.length - 1) % 2 === 0,
    "wire_length",
    "wire oracle is not lowercase even-length hex with one newline",
  );
  return Buffer.from(text.slice(0, -1), "hex");
}

function applyWireMutations(source, mutations) {
  let value = Buffer.from(source);
  for (const mutation of mutations) {
    if (mutation.op === "replace_byte") {
      requireWire(
        Number.isInteger(mutation.offset) &&
          mutation.offset >= 0 &&
          mutation.offset < value.length &&
          /^[0-9a-fA-F]{2}$/.test(mutation.value_hex),
        "wire_length",
        "invalid replace-byte mutation",
      );
      value[mutation.offset] = Number.parseInt(mutation.value_hex, 16);
    } else if (mutation.op === "replace_u32be") {
      requireWire(
        Number.isInteger(mutation.offset) &&
          mutation.offset >= 0 &&
          mutation.offset + 4 <= value.length,
        "wire_length",
        "invalid u32 mutation",
      );
      value.writeUInt32BE(mutation.value, mutation.offset);
    } else if (mutation.op === "replace_u64be") {
      requireWire(
        Number.isInteger(mutation.offset) &&
          mutation.offset >= 0 &&
          mutation.offset + 8 <= value.length,
        "wire_length",
        "invalid u64 mutation",
      );
      value.writeBigUInt64BE(BigInt(mutation.value), mutation.offset);
    } else if (mutation.op === "append_hex") {
      requireWire(
        /^(?:[0-9a-fA-F]{2})+$/.test(mutation.value_hex),
        "wire_length",
        "invalid append mutation",
      );
      value = Buffer.concat([value, Buffer.from(mutation.value_hex, "hex")]);
    } else if (mutation.op === "truncate") {
      requireWire(
        Number.isInteger(mutation.byte_length) && mutation.byte_length >= 0,
        "wire_length",
        "invalid truncate mutation",
      );
      value = value.subarray(0, mutation.byte_length);
    } else {
      throw new RunnerFailure("wire_length", "unsupported wire mutation operation");
    }
  }
  return value;
}

export async function executeWireCase(fixture, fixtureCase) {
  try {
    const descriptor = parseJsonBytesStrict(await readFixtureFile(fixture, fixtureCase.input));
    let wire;
    if (fixtureCase.case_kind === "wire_bytes") {
      const header = parseJsonBytesStrict(
        await readFixtureInternal(fixture, descriptor.header_source),
      );
      requireWire(
        /^(?:[0-9a-fA-F]{2})+$/.test(descriptor.payload_hex),
        "wire_length",
        "payload is not valid hexadecimal",
      );
      wire = encodeWireFrame(header, Buffer.from(descriptor.payload_hex, "hex"));
      decodeWireFrame(fixture, wire);
      const output = Buffer.from(`${wire.toString("hex")}\n`, "utf8");
      return caseResult(fixtureCase.case_id, "accept", null, [
        { kind: "wire_bytes", byte_length: output.length, sha256: sha256Hex(output) },
      ]);
    }
    if (fixtureCase.case_kind !== "wire_mutation")
      throw new RunnerFailure("wire_length", "unsupported wire fixture case kind");
    wire = parseWireHex(await readFixtureInternal(fixture, descriptor.base));
    decodeWireFrame(fixture, applyWireMutations(wire, descriptor.mutations ?? []));
    return caseResult(fixtureCase.case_id, "accept");
  } catch (error) {
    return rejectedCaseResult(fixtureCase, error);
  }
}
